keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1121 @@
+  general mutating permissions on an object type.
+
+### Non-Goals
+
+* Changing permissions for existing fields. This feature should have no
+  observable changes unless configured by administrators.
+
+## Proposal
+
+Let's introduce the user stories before the design:
+
+### User Stories
+
+#### Story 1
+
+I wish to restrict an operator "supersafe" to write only:
+
+* labels with keys beginning with "super.safe.com/"
+* annotations with keys beginning with "super.safe.com/"
+* finalizers with the prefix "super.safe.com/"
+* conditions with the type "SuperSafe"
+
+#### Story 2
+
+I'm adding a new field to pod.spec, "triggerSelfDestructOnSecurityBreach".
+Ordinary users must not be able to set this to true. Only admins may set it to
+true.
+
+#### List of known fields wanting these features
+
+* .metadata (all objects)
+  * .metadata.labels (parameterized by key)
+  * .metadata.annotations (parameterized by key)
+  * .metadata.finalizers (parameterized by prefix)
+  * .metadata.liens (doesn't exist yet; will be parameterized)
+  * .metadata.ownerReferences (maybe)
+* .status
+  * .status.conditions (many objects but not all; parameterized by specific type)
+* .spec
+  * .spec.replicas (multiple objects)
+  * .spec.nodeName (pod only)
+  * .spec.schedulePaused (pod only; doesn't exist yet)
+
+### Design
+
+We will describe a system of additional authorization checks apiserver will
+perform on mutating requests. First we need to explain some background and state
+the rest of the requirements, otherwise the design offered will seem too
+complex.
+
+#### Authorization System Background and Requirements
+
+On start-up, apiserver constructs an "authorizor" which is an object that
+understands how to correctly make authorization checks (henceforth "authz
+checks"). Users / cluster administrators can configure these checks to be done
+in a number of different ways. RBAC is the most obvious choice, since it is
+provided along with Kubernetes, but many clusters are integrated with external
+authz systems (using plugins).
+
+People offering reasonably dangerous extensions / new fields need assurances
+that these things can be secured in any environment, and that they can offer
+instructions on doing so mostly independently of the backend authz system
+(because they may not even be able to identify all implementations).
+
+Therefore, __modifying a single authz system can't solve the problem__.
+
+An authz check can be thought of as sending structured data to an opaque system
+which answers yes or no. Essentially we can't change things about the system
+("opaque") because the system could comprise any number of implementations, many
+not even open-source. The data we can give the system includes:
+
+(TODO: I'm sure sig auth will want to correct my wording here)
+
+* The resource type (group, version, resource)
+* The object's locator (namespace if any, name, uid)
+* The actor's identity (username, group)
+* The verb...
```

Comment:
verb is a free-form string, the kube-apiserver has a few well-known verbs like update.

Reply:
Added words